woocommerce/woocommerce-ios
# Fix Crash When Dismissing Message Composer

Fixes #1361 

## Findings

I got lucky with this one. I noticed that the crash had something like this:

```
messageComposeViewController:didFinishWithResult: > retainWeakReference > smsComposeControllerCancelled >
Attempted to dereference garbage pointer 0x3fe5607c.
```

I searched for `messageComposeViewController:didFinishWithResult` and found that in [`OrderMessageComposer`](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewModels/Order%20Details/OrderMessageComposer.swift). I didn't understand why this would crash but luckily I remembered how things like this could happen in Objective-C. After further investigation, I found out that [`messageComposeDelegate`](https://developer.apple.com/documentation/messageui/mfmessagecomposeviewcontroller/1614070-messagecomposedelegate) is an `unowned` reference. This gave me an idea that _maybe_ the `OrderMessageComposer` was deallocated **before** the `MFMessageComposerViewController` was dismissed. 

And it looks like that might be the case here. The crash can be reproduced by these steps:

1. In the web, submit an order with a phone number (preferably yours).
1. In `OrdersViewController.sync(...)`, replace the `ServiceLocator.stores.dispatch(action)` line with this:

```swift
        DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(10)) {
            ServiceLocator.stores.dispatch(action)
        }
```
2. Run the app on a real device.
3. Navigate to Orders.
4. Pull to refresh. 
5. **Very quickly** navigate to an Order → View Billing Info → Tap phone number → Message.
6. Wait until 10 seconds have passed and the order list should have been refreshed. 
7. Cancel the SMS dialog or send a message.

The app should crash after that. This is because the owner of `OrderMessageComposer`, [`BillingInformationViewController`](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/Customer%20Section/Billing%20Information/BillingInformationViewController.swift) was dismissed. This deallocates `OrderMessageComposer` which is the delegate of `MFMessageComposerViewController` whose UI is **still present**. When the user submits or cancels the dialog, `MFMessageComposerViewController` will try to inform the `messageComposeDelegate` but since it no longer exists → 💥 

The `BillingInformationViewController` was dismissed during the pull-to-refresh because we delete all the orders during pull-to-refresh and when an order is deleted, the `OrderDetailsViewController` dismisses the topmost VC:

 https://github.com/woocommerce/woocommerce-ios/blob/789ff4347ebb2f153982cedc578587735b71b29a/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/OrderDetailsViewController.swift#L119-L126

I'm not really sure if that code is a good idea. I don't think it's a good experience to suddenly dismiss a view while the user is doing something with it. But I'm opting not to touch that for now.

## Solution

I initially added a `deinit` in the `OrderMessageComposer` like this:

```swift
    deinit {
        // If this class is dismissed while a `ComposeViewController` is active, a crash will
        // happen because `ComposeViewController.delegate` is an `unowned` reference to `self`.
        messageComposeViewController?.dismiss(animated: false, completion: nil)
    }
```

After testing it, I found that it's not a good experience to dismiss the SMS dialog **while** the user is not finished with it. So, I went with the singleton (ew) approach by moving `OrderMessageComposer` to a reusable service as part of `ServiceLocator`: [`MessageComposerPresenter`](https://github.com/woocommerce/woocommerce-ios/blob/issue/1361-fix-crash-in-sms-compose/WooCommerce/Classes/ServiceLocator/MessageComposerPresenter.swift). This class is kept forever so it should not be deallocated. And it can dismiss `MFMessageComposerViewController` after the user is done. 

## Testing

1. Follow the steps to reproduce using `develop`. 
2. Confirm the crash happens.
3. Follow the same steps using this PR's branch. 
4. Confirm the crash no longer happens.

## Reviewing 

- Only 1 reviewer is needed but anyone can review. 
- Please review that I did not accidentally remove any critical code when I replaced `OrderMessageComposer`.

## Submitter Checklist
